pkg/engine/lifecycle_test.go

```diff
@@ -343,11 +344,16 @@ func (p *TestPlan) GetTarget(snapshot *deploy.Snapshot) deploy.Target {
 }
 
 func (p *TestPlan) Run(t *testing.T, snapshot *deploy.Snapshot) *deploy.Snapshot {
-	project, target := p.GetProject(), p.GetTarget(snapshot)
-
+	project := p.GetProject()
+	snap := snapshot
 	for _, step := range p.Steps {
+		// note: it's really important that the preview and update operate on different snapshots.  the engine can and
+		// does mutate the snapshot in-place, even in previews, and sharing a snapshot between preview and update can
+		// cause state changes from the preview to persist even when doing an update.
 		if !step.SkipPreview {
-			_, err := step.Op.Run(project, target, p.Options, true, step.Validate)
+			previewSnap := CloneSnapshot(t, snap)
```

Comment:
This fixes an issue where the preview and update operations shared a snapshot, which doesn't work because the engine updates the snapshot in-place even when doing previews.